src/rgw/rgw_sal.cc

```diff
@@ -41,45 +49,40 @@ int RGWRadosUser::list_buckets(const string& marker, const string& end_marker,
   if (ret < 0)
     return ret;
 
+  buckets.clear();
   buckets.set_truncated(is_truncated);
   for (const auto& ent : ulist.get_buckets()) {
-    RGWRadosBucket *rb = new RGWRadosBucket(this->store, *this, ent.second);
+    RGWRadosBucket *rb = new RGWRadosBucket(this->store, ent.second, this);
     buckets.add(rb);
   }
 
   return 0;
 }
 
-RGWBucketList::~RGWBucketList()
-{
-  for (auto itr = buckets.begin(); itr != buckets.end(); itr++) {
-    delete itr->second;
-  }
-  buckets.clear();
-}
-
-RGWBucket* RGWRadosUser::add_bucket(rgw_bucket& bucket,
+RGWBucket* RGWRadosUser::create_bucket(rgw_bucket& bucket,
 				       ceph::real_time creation_time)
 {
   return NULL;
 }
 
-int RGWRadosUser::get_by_id(rgw_user id, optional_yield y)
+int RGWRadosUser::load_by_id(optional_yield y)
 
 {
-    return store->ctl()->user->get_info_by_uid(id, &info, y);
+    return store->ctl()->user->get_info_by_uid(info.user_id, &info, y);
 }
 
-RGWObject *RGWRadosBucket::create_object(const rgw_obj_key &key)
+RGWObject* RGWRadosStore::get_object(const rgw_obj_key& k)
 {
-  if (!object) {
-    object = new RGWRadosObject(store, key);
-  }
+  return new RGWRadosObject(this, k);
+}
 
-  return object;
+/* Placeholder */
+RGWObject *RGWRadosBucket::create_object(const rgw_obj_key &key)
+{
+  return nullptr;
 }
 
-int RGWRadosBucket::remove_bucket(bool delete_children, optional_yield y)
+int RGWRadosBucket::remove_bucket(bool delete_children, std::string prefix, std::string delimiter, optional_yield y)
```

Comment:
prefix and delimiter look unused here? also, when would you want a bucket deletion to only delete a subset of its objects?

Reply:
It's passed to abort_bucket_multiparts(), and apparently used by swift.  Maybe that bit could be left out of this API?  I'll think about it.